File: ios/cryptography/Wallet/WalletCryptoBoxImpl.swift
import Foundation
import CryptoKit

class WalletCryptoBoxImpl: WalletCryptoBox {

    var selfCryptoBox: CryptoBoxImpl = CryptoBoxImpl()

    func getPublicKey() -> Data {
        return selfCryptoBox.getPublicKey()
    }

    func buildSecretsTranslator(verifierPublicKey: Data) -> SecretsTranslator {
        let secureRandom = secureRandomData(count: CryptoConstants.INITIALISATION_VECTOR_LENGTH)
        let selfcipherPackage = selfCryptoBox.createCipherPackage(otherPublicKey: verifierPublicKey, senderInfo: CryptoConstants.WALLET_INFO, receiverInfo: CryptoConstants.VERIFIER_INFO, ivBytes: secureRandom)
      return SenderTransfersOwnershipOfData(cipherPackage: selfcipherPackage, initVector: secureRandom)
    }

    func secureRandomData(count: Int) -> Data {
        var bytes = [Int8](repeating: 0, count: count)
        let status = SecRandomCopyBytes(
            kSecRandomDefault,
            count,
            &bytes
        )
        if status == errSecSuccess {
            // Convert bytes to Data
            let data = Data(bytes: bytes, count: count)
            return data
        }
        else {
            return Data()
        }
    }
}

File: kotlin/android/src/main/java/io/mosip/tuvali/cryptography/WalletCryptoBoxImpl.java
package io.mosip.tuvali.cryptography;

import java.security.SecureRandom;

class WalletCryptoBoxImpl implements WalletCryptoBox {
    private final CryptoBox selfCryptoBox;
    private SecureRandom secureRandom;

    WalletCryptoBoxImpl(SecureRandom random) {
        this.selfCryptoBox = new CryptoBoxBuilder().setSecureRandomSeed(random).build();
        this.secureRandom = random;
    }

    @Override
    public byte[] publicKey() {
        return selfCryptoBox.getPublicKey();
    }

    @Override
    public SecretsTranslator buildSecretsTranslator(byte[] verifierPublicKey) {
        byte[] nonceBytes = new byte[CryptoBox.NONCE_LENGTH];
        secureRandom.nextBytes(nonceBytes);

        CipherPackage cipherPackage = selfCryptoBox.createCipherPackage(verifierPublicKey, KeyGenerator.WALLET_INFO, KeyGenerator.VERIFIER_INFO, nonceBytes);
        return new SenderTransfersOwnershipOfData(nonceBytes, cipherPackage);
    }
}
